File: 0077-combinations/0077-combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans = []
        def backtracking(idx, tmp):
            if len(tmp) == k:
                ans.append(tmp.copy())
                return
            
            for i in range(idx, n+1):
                tmp.append(i)
                backtracking(i + 1, tmp)
                tmp.pop()
        backtracking(1, [])
        return ans
            

File: 0077-combinations/0077-combinations.kt
class Solution {
    fun combine(n: Int, k: Int): List<List<Int>> {
        val answer = mutableListOf<List<Int>>()

        fun dfs(idx: Int, tmp: MutableList<Int>){
            if (tmp.size == k){
                answer.add(tmp.toList())
                return
            }

            for (i in idx until n+1){
                tmp.add(i)
                dfs(i+1, tmp)
                tmp.removeAt(tmp.size - 1)
            }   
        }
        
        dfs(1, mutableListOf())
        return answer
    }
}